packages/grafana-runtime/src/utils/queryResponse.ts

```diff
@@ -127,6 +127,36 @@ export function toDataQueryResponse(
   return rsp;
 }
 
+/**
+ * Data sources using api/ds/query to test data sources can use this function to
+ * handle errors and convert them to TestingStatus object.
+ *
+ * If possible, this should be avoided in favor of implementing /health endpoint
+ * and testing data source with DataSourceWithBackend.testDataSource()
+ *
+ * @returns {TestingStatus}
+ */
+export function toTestingStatus(err: FetchError): any {
+  const queryResponse = toDataQueryResponse(err);
+  // POST api/ds/query errors returned as { message: string, error: string } objects
+  if (queryResponse.error?.data?.message && queryResponse.error?.data?.error) {
+    return {
+      status: 'error',
+      message: queryResponse.error.data.message,
+      details: { message: queryResponse.error.data.error },
```

Comment:
Based on earlier suggestion the details might not be included if `queryResponse.error.data.error` is undefined?

Reply:
@marefr Good catch. Though, I think the end result would be the same because when `data.error` is not defined then `toTestingStatus` re-throws the error and the message is extracted from the error in `public/app/features/datasources/state/actions.ts`:

```typescript
const message = errMessage || data?.message || 'HTTP error ' + statusText;
```

I have changed it so when `data.error` is missing it's handled earlier, in `toTestingStatus`, it's probably more readable this way.